docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+
+```mermaid
+flowchart LR
+    probe(s3gw probe) == die request #1  ==> radosgw(radosgw)
+    == death notice #1  ==> probe
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: blue
+```
+
+```mermaid
+flowchart LR
+    radosgw(radosgw) == start notice #1  ==> probe(s3gw probe)
+    probe(s3gw probe) == collect restart event #1  ==> probe(s3gw probe)
+
+    linkStyle 0 stroke: green
+    linkStyle 1 stroke: blue
+```
+
+- The user queries the tool for the statistics
+
+```mermaid
+flowchart LR
+    user[curl] == query statistics  ==> probe(s3gw probe)
+    == statistics ==> user
+
+    linkStyle 0 stroke: green
+    linkStyle 1 stroke: blue
+```
+
+## Tested scenarios
+
+As previously said, we want to compute some statistics regarding the Kubernetes
+performances when restarting the `radosgw`'s POD.
+The `radosgw` code has been patched to accept a REST call from a client
+where the user can specify the way the `radosgw` will exit.
+Currently, 3 modes are possible:
+
+- `EXIT_0`
+- `EXIT_1`
+- `CORE_BY_SEG_FAULT`
+
+### EXIT-1, 10 measures
+
+```yaml
+  - mark: exit1
+    series:
+      - restart_id: 1
+        duration: 0
+      - restart_id: 2
+        duration: 15
+      - restart_id: 3
+        duration: 24
+      - restart_id: 4
+        duration: 42
+      - restart_id: 5
+        duration: 91
+      - restart_id: 6
+        duration: 174
+      - restart_id: 7
+        duration: 307
+      - restart_id: 8
+        duration: 302
+      - restart_id: 9
+        duration: 302
+      - restart_id: 10
+        duration: 304
+    time_unit: s
+```
+
+<h1 align="left"><img alt="chatterbox-logo" src="./assets/1node-10restarts-exit1.png"/>
+</h1>
+
+### EXIT-0, 10 measures
+
+```yaml
+  - mark: exit0
+    series:
+      - restart_id: 1
+        duration: 1
+      - restart_id: 2
+        duration: 13
+      - restart_id: 3
+        duration: 25
+      - restart_id: 4
+        duration: 49
+      - restart_id: 5
+        duration: 91
+      - restart_id: 6
+        duration: 161
+      - restart_id: 7
+        duration: 302
+      - restart_id: 8
+        duration: 304
+      - restart_id: 9
+        duration: 305
+      - restart_id: 10
+        duration: 308
+time_unit: s
+```
+
+<h1 align="left"><img alt="chatterbox-logo" src="./assets/1node-10restarts-exit0.png"/>
+</h1>
+
+### Considerations on the tests performed
+
+While the Kubernetes's behavior for `EXIT-1` was the expected one, the writer
+was expecting a different behavior for the `EXIT-0` test.
+
+It is reasonable for Kubernetes to put a POD whose process exits with a code
+different than zero into `CrashLoopBackoff` state; it is less obvious that the
+same state is applied also for a POD whose process exits "normally".
+A clarification: Kubernetes puts a POD exited with `0` into `Completed` state,
+but if that POD is rescheduled, the state becomes `CrashLoopBackoff` and thus
+the behavioral schema is the same as if the process run by the POD exited
+abnormally.
```

Comment:
I agree that from the point of view of _ReplicaSet_ and _Deployment_, that are meant to keep a Pod alive, the _CrashLoopBackoff_ state when the Pod keeps exiting can make sense.

The tests implemented here are measuring the time between two events occurring on `radosgw`'s Pod: 

1. the death
2. the start  

I don't think it is similar to the concept of liveness probe (?).
The liveness probe is basically an arbitrary check to see if a Pod must be considered alive or not.

Reply:
The time between the death and the start is determined by how long K8s takes to a) detect the Pod is dead, which is influenced by the liveness probe and b) the time K8s takes to attempt a restart, which is influenced by the backoff function.
Additionally, the startup may be delayed if the container image has to be fetched from a registry, but that case is uninteresting for us in this discussion.
Therefore, what you are measuring here is mostly the probe and the backoff, since the actual scheduling and startup is insignificant to those two factors.
Keeping in mind that the backoff function [may also include jitter](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Backoff), you'll notice that the restarts happen close-enough at the [documented 10s, 20s, 40s, ... intervals](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy) up until the backoff limit at five minutes is reached.

The liveness probe influences how long K8s takes to detect a dead Pod, because it can be configured to do different kinds of checks at different frequencies and tolerate different counts of failed checks before the probe considers the Pod dead.
E.g. if your liveness probe is to make a HTTP Get request against a `/health` endpoint at a frequency of once every three seconds and you tolerate up to three failed checks and you also tolerate the response to take up to 300ms, it will take almost ten seconds to detect a failed Pod.
Conversely if your probe is a simple TCP connect every second and you tolerate only one failed check and the timeout is set to 100ms, you will detect a dead Pod in well under two seconds.


The backoff function will introduce an additional wait time if a Pod fails repeatedly, increasing the wait time each time the Pod fails.
So after the first time the Pod dies, K8s will immediately restart it, but after the second time, K8s will first wait a few seconds, after the third time it will wait a few seconds more and then more and more.
This is a feature that prevents situations where mass-restarts of Pods (e.g. in the case of a node failure) lead to perpetual resource starvation. Eventually, the backoff function will cause the restart to wait long enough that a mass-restart event is distributed over time.

Lastly, there is also the startup probe. This is different from the liveness probe as it is a check dedicated to determine if the Pod has started. In cases where you have a significant startup time (e.g. databases that have to replay a write-ahead-log) this can be utilized to let the Pod run, but not consider it dead even if the liveness probe wouldn't succeed. Once the startup probe has determined that the Pod is started up, the liveness probe will be considered again.

Reply:
Let's keep in mind that these tests implemented for this research are a bit different; here we have a customized `radosgw` process that `pushes` the `death` and `start` event to the `probe` custom tool (with the patches implemented [here](https://github.com/giubacc/ceph/tree/ha-testing)).
I designed the tests this way because I wanted to precisely measure the point in time we can say the `radosgw` is ready to accept new connections from a client.
This approach is not the classic `polling` mode you can do with a k8s probe, but it is a `push` (artificially reversed) way to collect death and start events.